runtime/j9vm/java11vmi.c

```diff
@@ -348,13 +352,54 @@ trcModulesCreationPackage(J9VMThread *currentThread, J9Module *fromModule, const
 }
 
 static BOOLEAN
-addPackageDefinition(J9VMThread * currentThread, J9Module * fromModule, const char *package)
+addPackageDefinition(J9VMThread *currentThread, J9Module *fromModule, const char *package)
 {
-	J9ClassLoader * const classLoader = fromModule->classLoader;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions const *const vmFuncs = vm->internalVMFunctions;
+	J9ClassLoader *const classLoader = fromModule->classLoader;
 
 	BOOLEAN retval = FALSE;
 
-	J9Package * j9package = createPackage(currentThread, fromModule, package);
+	PORT_ACCESS_FROM_VMC(currentThread);
+#if defined(J9VM_OPT_SNAPSHOTS)
+	VMSNAPSHOTIMPLPORT_ACCESS_FROM_JAVAVM(vm);
+#endif /* defined(J9VM_OPT_SNAPSHOTS) */
+	J9Package *j9package = NULL;
+	J9UTF8 *packageName = NULL;
+	UDATA packageNameLength = strlen(package);
+	if (packageNameLength < J9VM_PACKAGE_NAME_BUFFER_LENGTH) {
+		UDATA packageNameJ9UTF8Size = packageNameLength + sizeof(J9UTF8) + 1; /* +1 for null-terminator. */
+#if defined(J9VM_OPT_SNAPSHOTS)
+		if (IS_SNAPSHOTTING_ENABLED(vm)) {
+			packageName = (J9UTF8 *)vmsnapshot_allocate_memory(packageNameJ9UTF8Size, OMRMEM_CATEGORY_VM);
+		} else
+#endif /* defined(J9VM_OPT_SNAPSHOTS) */
+		{
+			packageName = (J9UTF8 *)j9mem_allocate_memory(packageNameJ9UTF8Size, OMRMEM_CATEGORY_VM);
+		}
+		if (NULL == packageName) {
+			vmFuncs->setNativeOutOfMemoryError(currentThread, 0, 0);
+			return retval;
+		}
+		memcpy(J9UTF8_DATA(packageName), (void *)package, packageNameLength);
+		J9UTF8_DATA(packageName)[packageNameLength] = '\0';
+		J9UTF8_SET_LENGTH(packageName, (U_16)packageNameLength);
```

Comment:
The main other use of `addUTFNameToPackage` is for setting up queries to the package hash tables. Those names do not need to be persisted (nor properly `free`'d when freeing a `J9Package`) and thus they shouldn't be allocated with the `vmsnapshot_allocate_memory` when snapshotting is enabled.

Given that it is only on creation where `J9Package::packageName` needs to potentially be allocated in such a way that it can be persisted, and that is only done once in the code base, I figured that an inline version, like implemented here, was more appropriate.

Reply:
Ok, let's keep it as-is for now. My initial intent was to aim for a common allocation path to improve maintainability and code clarity, and rely upon the compiler to do the optimizations.